### PR TITLE
Handle broadcasts across dot and concat

### PR DESCRIPTION
**Pull request by pfultz2**
This improves the constant propagation for bert models. Larger batch size no longer use as large of constants.

### CODE REVIEW

**Review by umangyadav — COMMENTED** [review]
Looks good. Need to fix some tests in CI though. also Performance tests have `nan` and BERT is dropping in perf.

File: src/simplify_algebra.cpp [hunk 1]
```diff
@@ -388,12 +383,72 @@ struct find_inner_broadcast
...
+                auto axis = v.at("axis").to<std::size_t>() - naxes;
+                return m.insert_instruction(
+                    ins, make_op("broadcast", {{"axis", axis}, {"out_lens", lens}}), input);
+            }
+            assert(false);
+            return m.end();
+        };
+        auto a1        = insert_broadcast(a);
+        auto b1        = insert_broadcast(b);
+        auto dot       = m.insert_instruction(ins, make_op("dot"), a1, b1);
```
Comment (umangyadav) [inline src/simplify_algebra.cpp:438]: `dot` seems to be returning standard-shaped outputs in `compute_shape()`.  This transformation would make it return broadcasted output.

Comment (umangyadav) [reply src/simplify_algebra.cpp:438]: Not sure if this would add extra `contiguous` if consumers expect standard shapes.

**Comment by causten** [comment]
I tested this PR.  I saw a performance gain.  More interestingly it also significantly speed up the compile phase.

File: src/simplify_algebra.cpp (hunk 1 above)
Comment (CharlieL7) [reply src/simplify_algebra.cpp:438]: I think that would be a fine tradeoff in comparison to doing a larger gemm.

**Review by CharlieL7 — COMMENTED** [review]
Jenkins looks to be failling on py_backend for shufflenet_cpu.

File: src/simplify_algebra.cpp (hunk 1 above)
Comment (CharlieL7) [reply src/simplify_algebra.cpp:438]: Should be fine in comparison to doing a larger gemm.

**Review by TedThemistokleous — APPROVED** [review]

**Review by CharlieL7 — APPROVED** [review]

**Pull request merged**